jni/src/common/Foreign/JNI/Unsafe/Introspection.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+{-# OPTIONS_GHC -fno-warn-name-shadowing #-}
+
+module Foreign.JNI.Unsafe.Introspection
+ ( getClassName
+ , getSignatures
+ , toString
+ , isInstanceOf
+ , classGetNameMethod
+ ) where
+
+import Control.Monad (forM)
+import Data.Foldable (fold)
+import Data.Maybe (catMaybes)
+import Data.ByteString.Builder (toLazyByteString, word16LE)
+import Data.ByteString.Lazy (toStrict)
+import Data.Singletons
+import Data.Text (unpack)
+import Data.Text.Encoding (decodeUtf16LE)
+import Foreign.JNI.Unsafe.Internal
+import Foreign.JNI.Types
+import qualified Foreign.JNI.String as JNI
+import Foreign.Marshal.Array
+import Prelude hiding (String)
+
+-- | The "Class" class.
+kclass :: IO JClass
+kclass = findClass $ referenceTypeName $ sing @('Class "java.lang.Class")
+
+-- | The "Method" class
+kmethod :: IO JClass
+kmethod = findClass $ referenceTypeName $ sing @('Class "java.lang.reflect.Method")
+
+-- | Class.getMethods
+classGetMethodsMethod :: IO JMethodID
+classGetMethodsMethod = do
+  klass <- kclass
+  let sig = methodSignature [] (SArray $ sing @('Class "java.lang.reflect.Method"))
+  getMethodID klass (JNI.fromChars "getMethods") sig
+
+-- | Method.toString
+methodToStringMethod :: IO JMethodID
+methodToStringMethod = do
+  klass <- kmethod
+  let sig = methodSignature [] (sing @('Class "java.lang.String"))
+  getMethodID klass (JNI.fromChars "toString") sig
+
+-- | Method.getName
+methodGetNameMethod :: IO JMethodID
+methodGetNameMethod = do
+  klass <- kmethod
+  let sig = methodSignature [] (sing @('Class "java.lang.String"))
+  getMethodID klass (JNI.fromChars "getName") sig
+
+-- | Class.getName
+classGetNameMethod :: IO JMethodID
+classGetNameMethod = do
+  klass <- kclass
+  let sig = methodSignature [] (sing @('Class "java.lang.String"))
+  getMethodID klass (JNI.fromChars "getName") sig
+
+-- | Class.isInstance
+classIsInstanceMethod :: IO JMethodID
+classIsInstanceMethod = do
+  klass <- kclass
+  let sig = methodSignature [SomeSing $ sing @('Class "java.lang.Object")] (sing @('Prim "boolean"))
+  getMethodID klass (JNI.fromChars "isInstance") sig
+
+-- | @getSignatures c methodName@ yields the Java signatures of overloadings of
+-- methods called @methodName@ in class @c@.
+getSignatures :: JClass -> JNI.String -> IO [JNI.String]
+getSignatures c methodName = do
+  id <- classGetMethodsMethod
+  array :: JObjectArray <- unsafeCast <$> callObjectMethod c id []
+  l <- getArrayLength array
+  names <- forM [0 .. l - 1] $ \i -> do
+    ithObj :: JObject <- getObjectArrayElement array i
```

Comment:
You need to make sure the reference is deleted, at this level it could be done with:
```Haskell
bracket (getObjectArrayElement array i) deleteLocalRef $ \(ithObj :: JObject) ->
    ...
```

`bracket` would be necessary when creating the `array` reference above as well.